cxx/gendb.cc

```diff
@@ -0,0 +1,157 @@
+  }
+}
+
+// This function walks the class_path of the query, populates the global
+// reference_values table if necessary, and returns a sampled set of items
+// for the query relation.
+std::vector<int> GenDB::sample_entities_relation(
+    std::mt19937* prng, const std::string& class_name,
+    std::vector<std::string>::const_iterator class_path_start,
+    std::vector<std::string>::const_iterator class_path_end, int class_item) {
+  if (class_path_end - class_path_start == 1) {
+    // These are domains and we need to DFS-traverse the class's
+    // parents, similar to PCleanSchemaHelper::compute_domains_for.
+    return sample_class_ancestors(prng, class_name, class_item);
+  } else {
+    // These are noisy relation domains along the path from the latent cleanly-
+    // observed class to the record class.
+    std::string ref_field = *class_path_start;
+
+    // If the reference field isn't populated, sample a value from a CRP and
+    // add it to reference_values.
+    std::string ref_class =
+        std::get<ClassVar>(
+            schema.classes.at(class_name).vars.at(ref_field).spec)
+            .class_name;
+    if (!reference_values[class_name].contains(class_item)) {
+      sample_and_incorporate_reference(prng, class_name, class_item, ref_field,
+                                       ref_class);
+    }
+    std::vector<int> items = sample_entities_relation(
+        prng, ref_class, ++class_path_start, class_path_end,
+        reference_values[class_name][class_item][ref_field]);
+    items.push_back(class_item);
+    return items;
+  }
+}
+
+void GenDB::sample_and_incorporate_reference(std::mt19937* prng,
+                                             const std::string& class_name,
+                                             int class_item,
+                                             const std::string& ref_field,
+                                             const std::string& ref_class) {
+  int new_val = domain_crps[ref_class].sample(prng);
+
+  // Generate a unique ID for the sample and incorporate it into the
+  // domain CRP.
+  std::stringstream new_id_str;
+  new_id_str << class_name << class_item << ref_field;
+  int new_id = std::hash<std::string>{}(new_id_str.str());
+  reference_values[class_name][class_item][ref_field] = new_val;
+  domain_crps[ref_class].incorporate(new_id, new_val);
+}
+
+// Recursively incorporates samples into base relations.
+void GenDB::incorporate_query_relation(std::mt19937* prng,
+                                       const std::string& query_rel_name,
+                                       const T_items& items,
+                                       const ObservationVariant& value) {
+  RelationVariant query_rel = hirm->get_relation(query_rel_name);
+  T_items base_items = std::visit(
+      [&](auto nr) {
+        using T = typename std::remove_pointer_t<decltype(nr)>::ValueType;
+        auto noisy_rel = reinterpret_cast<NoisyRelation<T>*>(nr);
```

Comment:
Since this only works if query_rel stores a NoisyRelation, why not drop the std::visit and use std::get<NoisyRelation*>?

Reply:
This is visiting over RelationVariant (which contains Relation<bool>, Relation<double> etc.)

Reply:
That should be Relation(bool), Relation(double) with pointy brackets